File: java/search/BinarySearchTree.java
package search;

public class BinarySearchTree {
    public static void main(String[] args) {
        // Crear e insertar nodos en el árbol
        BinarySearchTree tree = new BinarySearchTree();
        tree.insert(50);
        tree.insert(30);
        tree.insert(70);
        tree.insert(20);
        tree.insert(40);
        tree.insert(60);
        tree.insert(80);
        
        // Existe
        int keyToFind = 40;
        boolean exists = tree.binarySearchTree(keyToFind);
        System.out.println("Existe el elemento " + keyToFind + "? " + exists);
        
        // No existe
        keyToFind = 90;
        exists = tree.binarySearchTree(keyToFind);
        System.out.println("Existe el elemento " + keyToFind + "? " + exists);
    }

    public BinarySearchTree() {
        root = null;
    }

    static class Node {
        int key;
        Node left, right;
        
        public Node(int item) {
            key = item;
            left = right = null;
        }
    }

    private Node root;
    
    public void insert(int key) {
        root = insertRec(root, key);
    }
    
    private Node insertRec(Node root, int key) {
        if (root == null) {
            root = new Node(key);
            return root;
        }
        
        if (key < root.key) {
            root.left = insertRec(root.left, key);
        } else if (key > root.key) {
            root.right = insertRec(root.right, key);
        }
        
        return root;
    }
    
    public boolean binarySearchTree(int key) {
        return searchRec(root, key);
    }
    
    private boolean searchRec(Node root, int key) {
        if (root == null) {
            return false;
        }
        
        if (root.key == key) {
            return true;
        }
        
        if (key < root.key) {
            return searchRec(root.left, key);
        }
        
        return searchRec(root.right, key);
    }
}

File: python/search/binary_search_tree.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None
    
    def insert(self, key):
        """Insertar un nodo en el árbol."""
        if self.root is None:
            self.root = Node(key)
            return
        
        self._insert_recursive(self.root, key)
    
    def _insert_recursive(self, node, key):
        """Inserción recursiva de un nodo."""
        if key < node.key:
            if node.left is None:
                node.left = Node(key)
            else:
                self._insert_recursive(node.left, key)
        elif key > node.key:
            if node.right is None:
                node.right = Node(key)
            else:
                self._insert_recursive(node.right, key)
    
    def search(self, key):
        """Buscar un valor en el árbol."""
        return self._search_recursive(self.root, key)
    
    def _search_recursive(self, node, key):
        """Búsqueda recursiva de un valor."""
        if node is None or node.key == key:
            return node is not None
        
        if key < node.key:
            return self._search_recursive(node.left, key)
        return self._search_recursive(node.right, key)

if __name__ == "__main__":
    # Crear e insertar nodos en el árbol
    tree = BinarySearchTree()
    tree.insert(50)
    tree.insert(30)
    tree.insert(70)
    tree.insert(20)
    tree.insert(40)
    tree.insert(60)
    tree.insert(80)
    
    # Existe
    key_to_find = 40
    exists = tree.search(key_to_find)
    print(f"¿Existe el elemento {key_to_find}? {exists}")
    
    # No existe
    key_to_find = 90
    exists = tree.search(key_to_find)
    print(f"¿Existe el elemento {key_to_find}? {exists}")
